### PR TITLE
Route lib / midi io / dupl time sel

### PR DESCRIPTION
**Pull request by molleweide**
Also adds insertEmptyItem

### CODE REVIEW

**Comment by gwatcha** [comment]
This issue has conflicts as well. Please resolve

**Comment by molleweide** [comment]
1. l have built the foundation for a routing lib which allows for pretty much anything. It is based on a special type of string that you type and this allows for one to specify any kinds of mult to mult tracks. You can use str.match to batch route many tracks based on target guids or search strings.
2. Duplicate timeseection, ie. repeat segments of an arrangement.
3. midi device selection by keyvommand. 
      I created another repo called reaper-jsfx where I started to work on a vim-midi-controller
      lib so that on can access lots of midi functionality fast. As an example if i switch between
      my split keyb and my real piano i want to have my personal vim settup for each devick.
      where i switch to arpeggiator, chords, trahspode etc.
      basically handle all midi device stuff with eel.

**Comment by molleweide** [comment]
I also tried to setup a new virtual keyboard mode for testing but i am not getting it to work.
I have done the foundation necessary to send midi messages but I don't get the mode to work
properly from the beginning I think. Could you please take a look at that and tell me what i am doing wrong. I am thinking that is maybe obvious to you what i am doing wrong.

**Comment by molleweide** [comment]
I can also tell you that with this route lib you never will have to use the mouse for routing.
however it might need some more improvements but this is getting quite good actually.

I am still using log statements to show data. I haven't learnt the gui yet so I am logging route info. I'll try to checkout the gui more now.

**Comment by molleweide** [comment]
oh an I also added a function for splitting track name up into regex_prefix and main name incase
one wants to dive into more structured naming of tracks etc

**Comment by molleweide** [comment]
I also see that you have made many improvements with the bindings. it is awesome to now that the gui is back. it reappeard after i pulled master now before making the pr

**Comment by molleweide** [comment]
i have found some minor bugs in the routing lib that i am fixing atm

**Comment by molleweide** [comment]
Hey I just realized I should probably have made the pr to development branch, or no??

**Comment by molleweide** [comment]
One more thing i realize now is that i included my fx lib. I am coming up with many more ideas for working w fx bow that i am looking at sidechaining fx

**Comment by molleweide** [comment]
I added fx lib as well!

**Comment by molleweide** [comment]
something that i want to add i an fx_mode where one can move up/down and navigate params fast. I think I have some
good ideas on how to solve some of the annoying ux obtacles that would require mouse normally.

**Comment by molleweide** [comment]
i have a tiny fix that i am doing before you if you merge

**Comment by gwatcha** [comment]
Thankx @molleweide! I'll have more time to really delve into this in a couple of days

**Comment by molleweide** [comment]
Im looking forward to your feedback

**Comment by molleweide** [comment]
l am working on a scheme so that one can easilly have the same midi_keys layout between
regular laptop keyboard, ergodox/qmk. i am trying to set this up now in the jsfx project.
my goal is to be able to seemlessly be able to move between my laptop, ergodox, (i'd like to get a pair of dactyl manu also) without every having to rethink muscle memory.

i  have come up with some nice ideas. but for now since i cannot get key-release events I only
create fixed length notes on laptop with vkb. but this works fine for now.

as i  mentionned in another issue, i think, one interesting thing to do would be to suspend repear
keys when one is entering virtual keyboard mode somehow and let events pass through to the
built-in vkb. and only use eg. ESC to go back to normal mode. this would be the easiest i am imagining but i am unsure of how rk could be "turned off" during vkb mode.
on my ergodox this is not necessary since i have created a midi layer.

my idea above is to come up with some kind of generalized lyout that is easy to mirror on laptop.
tell me what you think later.

**Comment by molleweide** [comment]
also please checkout the pr but i am still finishing up my temporary custom vkb for laptop with US international qwerty

**Comment by molleweide** [comment]
i am going to add my ergodox midi layout in the reaper-jsfx repo so that you can understand how it is mirrored to my laptop keyboard.

**Comment by molleweide** [comment]
just FYI if you are looking at the `library/io_devices.lua` and `library/midi.lua` they
are sort of work in progress because I haven't finished the latest version of the `reaper-jsfx`
project but hopefully you will be able to see sort of where I am going.

**Comment by molleweide** [comment]
i am surprised how fast/responsive vkb is when I access it through rk. not much percieved latency at all!

**Comment by molleweide** [comment]
If you look at https://github.com/molleweide/reaper-jsfx/tree/molleweide you can see how
far i have gotten so far. In my first version every thing worked but now there is something that is not quite right so I have to fix that. But here you will be able to see where I am trying to go sort of. 
I am not sure what the error is. I jsfx is a bit annoying because it is much harder to debug logic than with lua. i am still trying to figure out a way of `how one could 'log' stuff?` in jsfx.
but when everything works it is very nice.

**Comment by gwatcha** [comment]
Hey @molleweide I'm happy you're having fun working on Reaper-Keys. I took a look through the entire PR (sorry it took me so long  😅).

I'm still a bit confused about what all this does, I don't use many MIDI devices myself. Could you explain a user case maybe?

**Comment by gwatcha** [comment]
It might be that it will have to live on your fork of the repo, and you could point the people that need it there through the Reaper-Forum or something of the like. This is just because the functionality is rather advanced for most users, and if I'm understanding the use case right, only a few of the users of Reaper Keys would be using it, so it's best not to include it I believe. I say this because of there are already many menu items already, and it would take a lot of time to really get into this PR and do a back & forth to get it into a mergeable state - right now the code sticks out compared to the code style of the rest of the repository, so there would have to be a lot of changes made.

Nevertheless, it's great you managed to get RK working for your usability issue with your MIDI devices, I hope more people learn about it. I wish you luck on your coding adventures :)

**Comment by molleweide** [comment]
yo!
Thanks for checking out my code. I hope you see some of it useful. I have tried to keep things as well structured as i can.
Please do give me feedback and tell me if I am making some beginner mistakes.

Yeah you don't have to include reaper-jsfx. It was mostly just for showing you how I connect rk to jsfx AND I think that the project is in its very early stage. It might become good but I have to work more on it.

**Comment by molleweide** [comment]
You have to tell me what you think of the routing stuff. it has to be tested more but i think that it is a great foundation. all though it can probably be refactored and bit more compacted.

**Comment by gwatcha** [comment]
Okay I'll leave some feedback for you! I'm not an expert software engineer but I had a co-op for a year and had to handle a lot of pull request feedback before I managed to push my first code, so I could try to do the same here.

**Comment by gwatcha** [comment]
If you really want to learn to code I think your best bet is to find something similar to a co-op so you can work in a team with more senior software people and get a lot of feedback.

**Comment by molleweide** [comment]
yeah i mean just tell me if you think i am being stupid or something. :D

File: definitions/defaults/actions.lua [hunk 1]
```diff
@@ -2,601 +2,676 @@ local lib = require('library')
...
+    "UnselectItems",
+    "SelectAllItemsInCurrentTimeSel",
+    lib.segments.insertSpaceAtEditCursorFromTimeSelection,
+    lib.segments.repeatShiftAllItemsInTimeSelectionByTrackByTimeSel
+  },
+
+  -- switch easy between devices
+  --
+  -- i included this here to show you mike
+  -- but this list of devices is ofc different for everybody.
```
Comment (gwatcha) [inline definitions/defaults/actions.lua:664]: it's best not to make conversation through code with PR's, leave that for github, 
in general, comments should be avoided at all costs in code because of this key principle: 'write code so that it does not need documentation'

the code is the docs. if you need to explain something, creating a new function with a descriptive name, or adding explanatory variables goes a long way

File: definitions/defaults/actions.lua [hunk 2]
```diff
@@ -2,601 +2,676 @@ local lib = require('library')
...
+  VKBSendMidi_G_2 = lib.midi.sendMidiNote_G_2,
+  VKBSendMidi_Fs2 = lib.midi.sendMidiNote_Fs2,
+  VKBSendMidi_F_2 = lib.midi.sendMidiNote_F_2,
+  VKBSendMidi_E_2 = lib.midi.sendMidiNote_E_2,
+  VKBSendMidi_Ds2 = lib.midi.sendMidiNote_Ds2,
+  VKBSendMidi_D_2 = lib.midi.sendMidiNote_D_2,
+  VKBSendMidi_Cs2 = lib.midi.sendMidiNote_Cs2,
+  VKBSendMidi_C_2 = lib.midi.sendMidiNote_C_2,
+
+  SelectAllItemsInCurrentTimeSel = 40717,
```
Comment (gwatcha) [inline definitions/defaults/actions.lua:648]: This section is very different from all the rest above, and it's best to keep things consistent when possible

File: definitions/config.lua [hunk 3]
```diff
@@ -13,6 +13,8 @@ return {
     'command',
     'operator',
     'meta_command',
-  }
-}
+  },
 
+  -- create you custom name prefix
```
Comment (gwatcha) [inline definitions/config.lua:18]: this is a user file and a user would not understand this at first glance, a more descriptive variable name may help

File: definitions/defaults/global.lua [hunk 4]
```diff
@@ -93,276 +95,298 @@ return {
...
+            ["4"] = "RecallSnapshot4",
+            ["5"] = "RecallSnapshot5",
+            ["6"] = "RecallSnapshot6",
+            ["7"] = "RecallSnapshot7",
+            ["8"] = "RecallSnapshot8",
+            ["9"] = "RecallSnapshot9",
+        }},
     }},
     [","] = {"+options", {
-               ["p"] = "TogglePlaybackPreroll",
```
Comment (gwatcha) [inline definitions/defaults/global.lua:116]: the indentation is meant to be consistent across the repository, so altering indentation levels is not good code quality

**Comment by molleweide** [comment]
Great. I will fix everything you say!

File: definitions/defaults/virtual_keyboard.lua [hunk 5]
```diff
@@ -0,0 +1,36 @@
+-- todo
+--
+--  i believe that I should be able to test now?!
+--
+--  it is quite annoying that i feel so annoyed by sitng at the cmputer.
+--  but i really feel now that thingr are gong to get really nice.
+--
+--  i do believe that it hould be possible to do this machine learning hing
```
Comment (gwatcha) [inline definitions/defaults/virtual_keyboard.lua:8]: there's a lot of comments that shouldn't go here, but I see you're typing something for me. Machine learning in reaper sounds interesting actually! what did you have in mind? I'm really interested in the intersection of code and music as well. Recently I got into the VCV Rack environment and programmed some modules there, maybe you will like that program too. It's philosophy is very Unix like 'do one thing and do it well'.

**Comment by molleweide** [comment]
in some cases i wanted stuff that I wrote to stand out so that you could comment on it but i realize that it becomes obvious
since pr highlights changes etc.

File: definitions/io.lua [hunk 6]
```diff
@@ -0,0 +1,13 @@
+-- starting to learn how to handle devices in code.
+
+return {
+  --  rename to midi_in ??
+  midi = {
+    vkb = 'Virtual Midi Keyboard', -- virtual keyboard | default
+    qmk = 'Ergodox EZ',
+    roland = 'Roland %- RD Series %- Port 1' -- my roland grand 88 key
```
Comment (gwatcha) [inline definitions/io.lua:8]: the comments here are all personalized as well. 

What I do for software is I keep an org file in emacs with information about my task I'm doing, and it helps keep me organized and keep track of information I need.

File: definitions/routing.lua [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+local TRACK_INFO_MIDIFLAGS_DISABLED = 4177951
+local TRACK_INFO_CATEGORY_SEND = 0 -- send
+local TRACK_INFO_CATEGORY_RECIEVE = -1 -- send
+local TRACK_INFO_CATEGORY_HARDWARE = 1 -- send
+
+--//////////////////////////////////////////////////
+--
+-- ROUTING
+--
+--  neither of a and m | a and not m
```
Comment (gwatcha) [inline definitions/routing.lua:15]: there are many comments here as well. they are difficult to understand for me too. Another reason why comments are evil in general is that they tend to duplicate information, and can grow out of data when code changes, leading to great  confusion

File: definitions/routing.lua [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+--
+--  only m
+--      create midi send
+--
+--  a and m
+--      create both audio and m
+--
+--  ----------------------------
+--
+--  the param_name is not necessarily the input format ?!
```
Comment (gwatcha) [inline definitions/routing.lua:26]: if comments are to be written it's best to imagine you are writing for a 5 year old because that's simple information is more digestable

File: definitions/routing.lua [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+
+  new_params = {--[[ empty ]]},
+
+  default_params = {
+    -- ////////////////////////////////////////////////////////////////////
+    -- WHICH TYPE
+    -- ////////////
+
+    ["a"] = {
+      description = 'SOURCE CHAN | int, index, &1024=mono, -1 for none',
```
Comment (gwatcha) [inline definitions/routing.lua:115]: I'm not sure exactly what this does still, but I believe you have made a scheme for getting arbitrary user input in the middle of key sequences, cool!

File: definitions/defaults/virtual_keyboard.lua (hunk 5 above)
Comment (molleweide) [reply definitions/defaults/virtual_keyboard.lua:8]: Cool! Sometimes I kind of 'write my mind' in comments so I don't forget stuff. In my case I am a beginner at machine learning but I have friends who do it and I have taken some courses on MIR stuff. In engineering school I used some supercollider and in the future I have been thinking it would be cool to use reaper in conjunction with supercollider and apply some ML on top of it and connect it to some web app maybe. I don't know really but I think that the best way for learning is to apply it on what I love. Let's make stuff as cyber as possible! My goal is basically to just learn as much about how to handle stuff in code and commndline now. I am starting to get good at computers now and so I just want to do stuff to learn as much as possible.
I will check out VCV. I recall the name!!

File: internal/command/action_sequence_functions/global.lua [hunk 10]
```diff
@@ -79,5 +79,13 @@ return {
         runner.extendTimelineSelection(move_function, args)
       end
     },
-  }
+  },
+  vkb = {
```
Comment (gwatcha) [inline internal/command/action_sequence_functions/global.lua:83]: using non shortened variable names improves readability, so I would rename this to virtual_keyboard_binding

File: internal/custom_actions/custom_actions.lua [hunk 11]
```diff
@@ -57,4 +60,90 @@ function custom_actions.splitItemsAtTimeSelection()
...
+  local retval, nameout = reaper.GetMIDIInputName( dev_id, '' )
+
+  local enabled_device
+  for k,device_str in pairs(io.midi) do
+    if nameout:lower():match(device_str:lower()) then enabled_device = device_str end
+  end
+
+  if enabled_device == nil then return end
+  if enabled_device == io.midi.vkb then
+    fx.setParamForFxAtIndex(tr, 0, 1, 0, true) -- set device
```
Comment (gwatcha) [inline internal/custom_actions/custom_actions.lua:108]: instead of comments to describe the functions here, I would create a variable or function to explain what is happening

File: internal/custom_actions/custom_actions.lua [hunk 12]
```diff
@@ -57,4 +60,90 @@ function custom_actions.splitItemsAtTimeSelection()
...
+  end
+
+  if enabled_device == io.midi.roland then
+    fx.setParamForFxAtIndex(tr, 0, 1, 2, true) -- set device
+    fx.setParamForFxAtIndex(tr, 0, 2, 6, true) -- set mode
+  end
+end
+
+
+function custom_actions.setupMidiInputPreProcessorOnSelTrks()
```
Comment (gwatcha) [inline internal/custom_actions/custom_actions.lua:124]: Trks is shortened here too, best to say Tracks, that would be consistent with the rest of the repository too

File: definitions/routing.lua (hunk 7 above)
Comment (molleweide) [reply definitions/routing.lua:15]: i have to clean up that file. it is way to messy!

File: definitions/routing.lua (hunk 9 above)
Comment (molleweide) [reply definitions/routing.lua:115]: yeah I hope this can be a good foundation for doing all route stuff with keys. and i think that with your skills it
could turn out really great in the end.

Comment (molleweide) [reply definitions/routing.lua:115]: In rlib string is where i filter string for route params. I am excited to hear what u think of this. Maybe you can come up with an even better way of doing stuff but a lot of the hwavylifting i hope is done. I just pray to good the code is somewhat readable so that you can test it.

Comment (molleweide) [reply definitions/routing.lua:115]: Now I have been trying to make action that setsup automatic sidechain routes based on track names. Eg send all kick drums to ghost kick track and then sidechain all tracks that begin with pattern to ghost kick. This kind of auto route would allieviate a shitload of manual stuff. Allso i try to setup automatic submixes based on track names.

Comment (molleweide) [reply definitions/routing.lua:115]: You creates route in popup user input.
Think ex-commands in vim like :w :set xyz.

File: definitions/defaults/virtual_keyboard.lua (hunk 5 above)
Comment (molleweide) [reply definitions/defaults/virtual_keyboard.lua:8]: Also i’ve had repetitive strain stuff and so is sometimes just write random stuff to diagnose myself and that’s why you might see some incomprehensible comments  😁

File: internal/custom_actions/custom_actions.lua (hunk 12 above)
Comment (molleweide) [reply internal/custom_actions/custom_actions.lua:124]: Yes!

File: internal/command/action_sequence_functions/global.lua (hunk 10 above)
Comment (molleweide) [reply internal/command/action_sequence_functions/global.lua:83]: Agree!

**Comment by molleweide** [comment]
Yo check out the route docs and tell me if you understand it.

File: definitions/routing.lua (hunk 9 above)
Comment (gwatcha) [reply definitions/routing.lua:115]: I am not spending too much time on this project anymore so I won't be able to help implement any functionality. 

The routing system also feel a bit unnecessary to me, I think users would prefer to do routing manually because it is not a common operation and it is simpler for them that way. The definitions for routings are also complicated and may confuse users looking into the definitions folder in order to configure RK. I won't be merging this request for this reason. Feature bloat is also a real thing!

Comment (gwatcha) [reply definitions/routing.lua:115]: I won't be merging this as it is right now for this reason. I may not merge it in the future after a lot of refactoring as well, so don't count on that! 

Thanks for your passion about Reaper Keys though, it's much appreciated to have someone else like this project as much as you do. If you would still like to help, I recommend trying to tackle some user issues like the confusion at the start and the completion window focus issue.

Comment (molleweide) [reply definitions/routing.lua:115]: Why are you not working on it so much anymore?
Yes I am going to look into those issues. I have to see how I can add them to my todo list.

**Comment by molleweide** [comment]
I am trying to learn linux atm. i am using virtualbox to test my configs on lots of linuxes.
it's gonna be fun to try reaper on linux!! i also just installed a program called kmonad for keys and mirroring my qmk. have you used it?

**Comment by gwatcha** [comment]
Hey I've been sick for a while now :face_with_thermometer: I'll get back to you when I feel better :smile:

**Comment by gwatcha** [comment]
> I am trying to learn linux atm. i am using virtualbox to test my configs on lots of linuxes.
> it's gonna be fun to try reaper on linux!! i also just installed a program called kmonad for keys and mirroring my qmk. have you used it?

I have not used it. Good luck! The biggest issue I had with reaper on Linux is getting all the vst's to work

File: definitions/routing.lua (hunk 9 above)
Comment (gwatcha) [reply definitions/routing.lua:115]: I'm developing a vision right now for a music interface project that is something like a Modular Digital Audio Workstation + a controller for it, which can optimally replace the function of the DAW and stay more true to music, real-time expression, and visions, and require less ego mental models to operate.

Comment (gwatcha) [reply definitions/routing.lua:115]: I bet you would be interested seeing as how you love coding and music as much as I do.

Comment (gwatcha) [reply definitions/routing.lua:115]: It's still in a watery intuition visionary stage in the pipeline of creation, so I've been holding off on sharing it for a while now. Here's a [demo](https://www.youtube.com/watch?v=FpHICKVpKT4) of an earlier version of the prototype for the controller I made which you may like.

Comment (gwatcha) [reply definitions/routing.lua:115]: So I've been really excited about this project for a while now and haven't been spending a lot of time on Reaper Keys anymore.

**Comment by molleweide** [comment]
> Hey I've been sick for a while now :face_with_thermometer: I'll get back to you when I feel better :smile: 
> 
> 

No worries. I hope you get well soon 👊

**Comment by molleweide** [comment]
yo I think that I am ready to take on spacemacs soon. lately I have been checking out a lot of functional programming stuff and I am realizing that neovim is written in lua which is imperative and emacs is written in lisp which is functional. A couple of days ago I played around with virtualbox to learn how to install my dotfiles on linux and I also decided to upgrade some tools. I started using nyxt web browser, also written in lisp, and I also started using kmonad for keymappings n stuff, which is written in haskell, and so I am starting to shift my world a bit more towards functional stuff. I am really looking forward now to learning spacemacs and see how it differs to neovim. However, I also adopted a guys dotfiles whose configs were highly geared towards emacs so I have the really nice spacebar legend.
On macos you can do automation stuff with hammerspoon and this allows for pretty much Lua programming everywhere. And now taking on some functional stuff is exciting. Btw, kmonad was amazing. Now I have been using it for a couple of days and my macbook pro feels exactly like my ergodox now with identical layout.

File: definitions/routing.lua (hunk 9 above)
Comment (molleweide) [reply definitions/routing.lua:115]: Dude I missed these comments lol. It's getting a bit cluttered.

Allright I see! I am going to checkout the demo. So could this program also be used together with reaper or is it like a full replacement sort of?

Comment (molleweide) [reply definitions/routing.lua:115]: Or would it make sense to reuse reaper keys to give you a modal keybingins system for switching between blocks etc?

Comment (molleweide) [reply definitions/routing.lua:115]: Would this be a more nische kind of app compared to reaper that woul maybe be more generalist kind of? I am not sure if I am at the level where I can even ask question about when and why to build a music app from scratch? I mean from my perspective it looks a little bit like a new gui that could be layered on top of reaper's audio engine or am I wrong?
Could I make drum beats with this as an example? I am probably asking some stupid questions since I don't know your full vision.

Comment (molleweide) [reply definitions/routing.lua:115]: could i use this instrument to record into reaper or would I do a full production with emergem blocks? What language do your write in?

Comment (molleweide) [reply definitions/routing.lua:115]: would this app be geared toward a particular genre? is it a midi instrument on it's own or do I have to work together with real instruments or midi instruments?

Comment (molleweide) [reply definitions/routing.lua:115]: What problems does this app solve that reaper doesn't?

**Comment by molleweide** [comment]
have you seen this https://forum.cockos.com/showthread.php?t=250419
I'm not really sure how it works but it seems to be a "big deal" according to many in the thread.

File: definitions/routing.lua (hunk 9 above)
Comment (gwatcha) [reply definitions/routing.lua:115]: Hey I'm working on videos to communicate the concepts, I'll share with you when I'm done. There's a lot of content.

Comment (molleweide) [reply definitions/routing.lua:115]: holy shit. cool!

**Comment by gwatcha** [comment]
I'm closing this to close this because I decided not to merge it.

**Comment by molleweide** [comment]
great! I'll create a new PR in the future when I have worked more on this.

**Comment by molleweide** [comment]
Hi Mike,

I hope you're doing well!

How is the music project going that you told me about? I have
been thinking about you from time to time now and so I wanted to see
how everything is going? Have you been coding a lot recently or
are you doing other stuff? 

Would you mind also setting up a discussions area for reaper-keys?

// Hjalmar

**Comment by gwatcha** [comment]
Hey! Do you have a telegram?

**Comment by molleweide** [comment]
Yes! What is your username on there? I'll try to find you.

**Comment by gwatcha** [comment]
@gwatcha
